# Проект: Секреты Темнолесья

* Автор: Филимонова Мария
* Дата: 14.01.2026

### Цели и задачи проекта

В этой тетрадке используем датасет `new_games.csv`, который содержит информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр.
Необходимо провести комплексный анализ рынка компьютерных игр для выявления ключевых тенденций, факторов успешности и рыночных закономерностей, которые могут быть использованы для оптимизации стратегий разработки и продвижения игр.

### Описание данных

В проекте будут использованы данные датасета `new_games.csv` с таким описанием:

- `Name` — название игры.
- `Platform` — название платформы.
- `Year of Release` — год выпуска игры.
- `Genre` — жанр игры.
- `NA sales` — продажи в Северной Америке (в миллионах проданных копий).
- `EU sales` — продажи в Европе (в миллионах проданных копий). 
- `JP sales` — продажи в Японии (в миллионах проданных копий).
- `Other sales` — продажи в других странах (в миллионах проданных копий).
- `Critic Score` — оценка критиков (от 0 до 100).
- `User Score` — оценка пользователей (от 0 до 10).
- `Rating` — рейтинг организации ESRB.

### Содержимое проекта
1. Загрузка и знакомство с данными
2. Проверка ошибок в данных и их предобработка
3. Фильтрация данных
4. Категоризация данных
5. Итоговый вывод

## 1. Загрузка и знакомство с данными:

In [1]:
# Импортируем библиотеку pandas
import pandas as pd

In [2]:
# Выгружаем данные из датасета new_games.csv в датафрейм df
try:
    df = pd.read_csv('/datasets/new_games.csv')
except FileNotFoundError:
    df = pd.read_csv('C:/Users/Huawei/OneDrive/Рабочий стол/Проекты/Секреты Темнолесья/new_games.csv')

In [3]:
# Выводим информацию о датафрейме
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [4]:
# Выводим первые строки датафрейма на экран
df.head()

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


Датасет `new_games.csv` содержит 11 столбцов и 16956 строк, в которых представлена информация о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр. 

Изучим типы данных и их корректность:
- **Числовые значения с плавающей запятой (float64).** Четыре столбца имеют тип данных `float64`:
    - `NA sales` и `Other sales` содержат информацию о продажах (в Северной Америке и в других странах в миллионах проданных копий), что логично для чисел с плавающей точкой. Здесь тип данных `float64` подходит.
    - `Year of Release` и `Critic Score` содержат информацию о годе выпуске игры и оценку критиков (от 0 до 100). Для таких данных рекомендуется использовать тип `int64`
- **Строковые данные (object).** Семь столбцов имеют тип данных `object`:
    - `Name`, `Platform`, `Genre` и `Rating` содержат строковую информацию (название игры, название платформы, жанр игры и рейтинг организации ESRB), что логично для текстовых данных. Здесь тип данных `object` подходит.
    - `EU sales` и `JP sales` содержат информацию о продажах (в Европе и в Японии в милллионах проданных копий). Для таких данных рекомендуется использовать тип `float64`
    - `User score` содержит информацию обоценке пользователей (от 0 до). Для таких данных рекомендуется использовать тип `float64`

После анализа типов данных видно, что больше половины столбцов корректно представлены. Однако для оптимизации можно тип данных некоторых столбцов необходимо преобразовть

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

In [5]:
# Выводим на экран названия всех столбцов датафрейма
df.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [6]:
# Приведем названия всех столбцов к snake_case
df.columns = df.columns.str.lower().str.replace(' ', '_')

In [7]:
# Выводим результат
df.columns

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')

### 2.2. Типы данных

In [8]:
# Выводим информацию о типах данных в датафрейме
df.dtypes

name                object
platform            object
year_of_release    float64
genre               object
na_sales           float64
eu_sales            object
jp_sales            object
other_sales        float64
critic_score       float64
user_score          object
rating              object
dtype: object

Выше показано, что `pandas` 'посчитал', столбцы `eu_sales`, `jp_sales` и `user_score` строковыми, а `year_of_release` и `critic_score` - вещественными числами

In [9]:
# Преобразуем к числовому типу данных с заменой пропусков на Nan
# Преобразуем eu_sales и jp_sales в числовой тип
df['eu_sales'] = pd.to_numeric(df['eu_sales'], errors='coerce')
df['jp_sales'] = pd.to_numeric(df['jp_sales'], errors='coerce')

# Преобразуем user_score в числовой тип
df['user_score'] = pd.to_numeric(df['user_score'], errors = 'coerce')

# Приводим к типу данных Int64
df['year_of_release'] = pd.to_numeric(df['year_of_release'], errors='coerce')
df['year_of_release'] = df['year_of_release'].astype('Int16')

In [10]:
# Проверим преобразованные типы данных
df.dtypes

name                object
platform            object
year_of_release      Int16
genre               object
na_sales           float64
eu_sales           float64
jp_sales           float64
other_sales        float64
critic_score       float64
user_score         float64
rating              object
dtype: object

##### **Возможные причины некорректных типов данных:**
- В столбцах могут встречаться строковые значения или они могут содержать пробелы, запятые и другие нечисловые символы
- Изначально данные были некорректно записаны

### 2.3. Наличие пропусков в данных

In [11]:
# Выводим количество пропущенных строк в датафрейме
df.isna().sum()

name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8714
user_score         9268
rating             6871
dtype: int64

In [12]:
# Подсчитываем процент строк с пропусками
df.isna().sum() / len(df) * 100

name                0.011795
platform            0.000000
year_of_release     1.621845
genre               0.011795
na_sales            0.000000
eu_sales            0.035386
jp_sales            0.023590
other_sales         0.000000
critic_score       51.391838
user_score         54.659118
rating             40.522529
dtype: float64

В данных наблюдаются пропуски в следующих столбцах:
- `name`: в 2 строках (<1% данных) отсутсвует информация о названии игры. Минимальное количество пропусков, скорее всего, техническая ошибка
- `year_of_release`: в 275 строках (1.6% данных) отсутсвует информация о годе выпуска игры. Возможно, было переиздание игры, игра была анонсированна, но не выпущена или техническая ошибка
- `genre`: в 2 строках (<1% данных) отсутствует информация о жанре игры. Минимальное количество пропусков, скорее всего, техническая ошибка
- `eu_sales`: в 6 строках (<1% данных) отсутствует информация о продажах в Европе. Пропусков немного, возможно, техническая ошибка
- `jp_sales`: в 4 строках (<1% данных) отсутствует информация о продажах в Японии. Пропусков немного, возможно, техническая ошибка
- `critic_score`: в 8714 строках (51.4% данных) отсутствует информация об оценке критиков. Возможно, игра не была оценена критиками или техническая ошибка
- `user_score`: в 9268 строках (54.7% данных) отсутствует информация об оценках пользователей.  Возможно, игра не была оценена пользователями или техническая ошибка
- `rating`: в 6871 строках (40.5% данных) отсутствует информация о рейтинге организации ESRB. Возможно, игра не имеет рейтинга ESRB или техническая ошибка

##### Обработка пропущенных значений:
- `name`, `genre`, `year_of_release`, `eu_sales` и `jp_sales` - пропуски немногочисленны, и вероятнее, что они случайны (технические причины). Можно считать, что это пропуски типа MCAR
- `critic_score` и `user_score` - пропуски могут быть как случайны (технические причины), так и могут зависеть, например, от платформы и жанра. Можно считать, что это пропуски типа MAR
- `rating` - пропуски могут быть не случайны. Создадим отдельную категорию `Unknown`

In [13]:
# Удаляем пропуски в столбце name, так как пропусков очень мало
df = df.dropna(subset=['name'])

In [14]:
# Удаляем пропуски в столбце genre, так как пропусков очень мало
df = df.dropna(subset=['genre'])

In [15]:
# Удаляем пропуски в столбце year_of_release, так как пропуски составляют всего несколько процентов
df = df.dropna(subset=['year_of_release'])

In [16]:
# Проверим, сколько осталось строк после удаления пропусков
df.shape

(16679, 11)

**Обработка пропусков в колонках `eu_sales` и `jp_sales`**

In [17]:
# Выведем первые 5 строк с пропусками в колонке 'eu_sales'
missing_in_eu_sales = df.loc[df['eu_sales'].isna(), ['eu_sales']]
display(missing_in_eu_sales.head(5))

,eu_sales
446,NaN
802,NaN
1131,NaN
1132,NaN
1394,NaN


In [18]:
# Выведем первые 5 строк с пропусками в колонке 'jp_sales'
missing_in_jp_sales = df.loc[df['jp_sales'].isna(), ['jp_sales']]
display(missing_in_jp_sales.head())

,jp_sales
467,NaN
819,NaN
1379,NaN
4732,NaN


In [19]:
# Обработаем пропуски в колонках 'eu_sales' и 'jp_sales'
def pass_processing_1(df, coll_name_1):
    # Вычисляем среднее значение по платформе и году
    group_mean = df.groupby(['platform', 'year_of_release'])[coll_name_1].mean()
    # Среднее по платформе
    platform_mean = df.groupby('platform')[coll_name_1].mean()
    # Среднее по колонке
    coll_mean = df[coll_name_1].mean()

    def staff_row_1(r):
        # Если значение не пропущено, возвращаем его
        if not pd.isna(r[coll_name_1]):
            return r[coll_name_1]

        # Получаем платформу и год для текущей строки
        platform_value = r['platform']
        year_value = r['year_of_release']

        if not pd.isna(platform_value) and not pd.isna(year_value):
            # Ищем среднее значение для данной комбинации платформы и года
            mean_value = group_mean.loc[(platform_value, year_value)]
            if not pd.isna(mean_value):
                return mean_value
        
        if not pd.isna(platform_value):
            mean_value = platform_mean.loc[platform_value]
            if not pd.isna(mean_value):
                return mean_value
            
        return coll_mean
    
    # Применяем функцию к каждой строке
    df[coll_name_1] = df.apply(staff_row_1, axis=1)
    return df

# Заполняем пропуски в колонках с информацией о продахаж
sales_list = ['eu_sales', 'jp_sales']

for coll in sales_list:
    if coll in df.columns:
        df = pass_processing_1(df, coll)

In [20]:
# Выведем первые 5 строк с пропусками в колонке 'eu_sales' после замены
missing_in_eu_sales = df.loc[df['eu_sales'].isna(), ['eu_sales']]
display(missing_in_eu_sales.head(5))

,eu_sales


In [21]:
# Выведем первые 5 строк с пропусками в колонке 'jp_sales' после замены
missing_in_jp_sales = df.loc[df['jp_sales'].isna(), ['jp_sales']]
display(missing_in_jp_sales.head())

,jp_sales


**Обработка пропусков в колонках `critic_score` и `user_score`**

In [22]:
# Выведем первые 5 строк с пропусками в колонке 'critic_score'
missing_in_critic_score = df.loc[df['critic_score'].isna(), ['critic_score']]
display(missing_in_critic_score.head())

,critic_score
1,NaN
4,NaN
5,NaN
9,NaN
10,NaN


In [23]:
# Выведем первые 5 строк с пропусками в колонке 'user_score'
missing_in_user_score = df.loc[df['user_score'].isna(), ['user_score']]
display(missing_in_user_score.head())

,user_score
1,NaN
4,NaN
5,NaN
9,NaN
10,NaN


In [24]:
# Обработаем пропуски в 'critic_score' и 'user_score'
# Для 'critic_score' используем индикатор -1, так как -1 выходит за диапазон оценивания
df['critic_score'] = df['critic_score'].fillna(-1)

# Для 'user_score' используем индикатор -1, так как -1 выходит за диапазон оценивания
df['user_score'] = df['user_score'].fillna(-1)

In [25]:
# Выведем первые 5 строк с пропусками в колонке 'critic_score' после замены
missing_in_eu_sales = df.loc[df['critic_score'].isna(), ['critic_score']]
display(missing_in_eu_sales.head(5))

,critic_score


In [26]:
# Выведем первые 5 строк с пропусками в колонке 'user_score', после замены
missing_in_eu_sales = df.loc[df['user_score'].isna(), ['user_score']]
display(missing_in_eu_sales.head(5))

,user_score


**Обработка пропусков в колонкe `rating`**

In [27]:
# Выведем первые 5 строк с пропусками в колонке 'rating'
missing_in_rating = df.loc[df['rating'].isna(), ['rating']]
display(missing_in_rating.head())

,rating
1,NaN
4,NaN
5,NaN
9,NaN
10,NaN


In [28]:
# Выводим уникальные значения в колонке 'rating'
rating_unique = df['rating'].unique()
display(rating_unique)

# В результате видим, что пропущенные значения представлены только 'Nan'

array(['E', nan, 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP'], dtype=object)

In [29]:
# Удаляем лишние пробелы в начале и конце строк
df['rating'] = df['rating'].str.strip()

In [30]:
# Заменяем пропуски 'Nan' значением 'Unknown'
replacement = 'Unknown'
df['rating'] = df['rating'].fillna(replacement)

In [31]:
# Выводим строки с пропусками в колонуе 'rating' после замены
missing_in_rating_after = df.loc[df['rating'].isna(), ['rating']]
display(missing_in_rating_after)

,rating


Проверим то, как заполнены пропуски. Выведем информацию о датафрейме

In [32]:
# Выводим информацию о датафрейме
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16679 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16679 non-null  object 
 1   platform         16679 non-null  object 
 2   year_of_release  16679 non-null  Int16  
 3   genre            16679 non-null  object 
 4   na_sales         16679 non-null  float64
 5   eu_sales         16679 non-null  float64
 6   jp_sales         16679 non-null  float64
 7   other_sales      16679 non-null  float64
 8   critic_score     16679 non-null  float64
 9   user_score       16679 non-null  float64
 10  rating           16679 non-null  object 
dtypes: Int16(1), float64(6), object(4)
memory usage: 1.4+ MB


### 2.4. Явные и неявные дубликаты в данных

In [33]:
# Функция для вывода уникальных значений для каждого столбца ('genre', 'platform', 'rating', 'year_of_release')
def unique_categorical_data(columns_data):
    # Получаем уникальные значения
    print(columns_data.upper())
    unique_val = df[columns_data].unique()
    print(f'Количество уникальных значений: {len(unique_val)}')
    print(f'Уникальные значения: {unique_val}')
    print()


columns_data = ['genre', 'platform', 'rating', 'year_of_release']

for coll in columns_data:
    if coll in df.columns:
        unique_categorical_data(coll)

GENRE
Количество уникальных значений: 24
Уникальные значения: ['Sports' 'Platform' 'Racing' 'Role-Playing' 'Puzzle' 'Misc' 'Shooter'
 'Simulation' 'Action' 'Fighting' 'Adventure' 'Strategy' 'MISC'
 'ROLE-PLAYING' 'RACING' 'ACTION' 'SHOOTER' 'FIGHTING' 'SPORTS' 'PLATFORM'
 'ADVENTURE' 'SIMULATION' 'PUZZLE' 'STRATEGY']

PLATFORM
Количество уникальных значений: 31
Уникальные значения: ['Wii' 'NES' 'GB' 'DS' 'X360' 'PS3' 'PS2' 'SNES' 'GBA' 'PS4' '3DS' 'N64'
 'PS' 'XB' 'PC' '2600' 'PSP' 'XOne' 'WiiU' 'GC' 'GEN' 'DC' 'PSV' 'SAT'
 'SCD' 'WS' 'NG' 'TG16' '3DO' 'GG' 'PCFX']

RATING
Количество уникальных значений: 9
Уникальные значения: ['E' 'Unknown' 'M' 'T' 'E10+' 'K-A' 'AO' 'EC' 'RP']

YEAR_OF_RELEASE
Количество уникальных значений: 37
Уникальные значения: <IntegerArray>
[2006, 1985, 2008, 2009, 1996, 1989, 1984, 2005, 1999, 2007, 2010, 2013, 2004,
 1990, 1988, 2002, 2001, 2011, 1998, 2015, 2012, 2014, 1992, 1997, 1993, 1994,
 1982, 2016, 2003, 1986, 2000, 1995, 1991, 1981, 1987, 1980, 1983]


In [34]:
# Функция для поиска неявных дубликатов
def implicit_duplicates(columns_name):

    column_data = df[columns_name]

    original_unique = column_data.unique()
    original_count = len(original_unique)

    normalized= column_data.astype(str).str.lower().str.strip()
    normalized_unique = normalized.unique()
    normalized_count = len(normalized_unique)

    print(columns_name.upper())
    print(f'Оригинальные уникальные значения {original_count}:')
    print(f'{sorted(original_unique)}')

    print(f'Уникальные нормализованные значения {normalized_count}:')
    print(f'{sorted(normalized_unique)}')
    print()
    

    if original_count > normalized_count:
        print(f'Вывод: {original_count - normalized_count} - количество неявных дубликатов')
    else:
        print('Вывод: Неявных дубликатов нет')

    print()
    print()


columns_name = ['genre', 'platform', 'rating']

for coll in columns_name:
    if coll in df.columns:
        implicit_duplicates(coll)

GENRE
Оригинальные уникальные значения 24:
['ACTION', 'ADVENTURE', 'Action', 'Adventure', 'FIGHTING', 'Fighting', 'MISC', 'Misc', 'PLATFORM', 'PUZZLE', 'Platform', 'Puzzle', 'RACING', 'ROLE-PLAYING', 'Racing', 'Role-Playing', 'SHOOTER', 'SIMULATION', 'SPORTS', 'STRATEGY', 'Shooter', 'Simulation', 'Sports', 'Strategy']
Уникальные нормализованные значения 12:
['action', 'adventure', 'fighting', 'misc', 'platform', 'puzzle', 'racing', 'role-playing', 'shooter', 'simulation', 'sports', 'strategy']

Вывод: 12 - количество неявных дубликатов


PLATFORM
Оригинальные уникальные значения 31:
['2600', '3DO', '3DS', 'DC', 'DS', 'GB', 'GBA', 'GC', 'GEN', 'GG', 'N64', 'NES', 'NG', 'PC', 'PCFX', 'PS', 'PS2', 'PS3', 'PS4', 'PSP', 'PSV', 'SAT', 'SCD', 'SNES', 'TG16', 'WS', 'Wii', 'WiiU', 'X360', 'XB', 'XOne']
Уникальные нормализованные значения 31:
['2600', '3do', '3ds', 'dc', 'ds', 'gb', 'gba', 'gc', 'gen', 'gg', 'n64', 'nes', 'ng', 'pc', 'pcfx', 'ps', 'ps2', 'ps3', 'ps4', 'psp', 'psv', 'sat', 'scd',

In [35]:
# Проверяем значения в колонках 'genre' и 'rating' до преобразования
df[['genre', 'rating']].head()

,genre,rating
0,Sports,E
1,Platform,Unknown
2,Racing,E
3,Sports,E
4,Role-Playing,Unknown


In [36]:
# Приводим значения в колонке 'genre' к нижнему регистру
df['genre'] = df['genre'].str.lower()

# Приводим значения в колонке 'rating' к верхнему регистру
df['rating'] = df['rating'].str.upper()

In [37]:
# Проверяем значения в колонках 'genre' и 'rating' после преобразования
df[['genre', 'rating']].head()

,genre,rating
0,sports,E
1,platform,UNKNOWN
2,racing,E
3,sports,E
4,role-playing,UNKNOWN


**Проверка на наличие полностью дублирующих строк**

In [38]:
# Вывод явных дубликатов
duplicates = df[df.duplicated()]
print("Дублирующиеся строки:")
print(duplicates)

Дублирующиеся строки:
                                                name platform  \
268                            Batman: Arkham Asylum      PS3   
368                 James Bond 007: Agent Under Fire      PS2   
717                            God of War: Ascension      PS3   
823                                Wipeout: The Game      Wii   
848                  Rayman Raving Rabbids: TV Party      Wii   
...                                              ...      ...   
16671        Fullmetal Alchemist: Prince of the Dawn      Wii   
16753                                      Routes PE      PS2   
16799                            Transformers: Prime      Wii   
16912  Metal Gear Solid V: The Definitive Experience     XOne   
16940                          The Longest 5 Minutes      PSV   

       year_of_release      genre  na_sales  eu_sales  jp_sales  other_sales  \
268               2009     action      2.24      1.31      0.07         0.61   
368               2001    shooter    

In [39]:
# Сумма явных дубликатов
df.duplicated().sum()

np.int64(235)

In [40]:
# Удаляем полные дубликаты
df.drop_duplicates(inplace=True)

# Вывод размера датафрейма после удаления полных дубликатов
display(df.shape)

(16444, 11)

**Промежуточные выводы**

**1. Явные дубликаты:**

- Полные дубликаты: найдено 209 строк

**2. Неявные дубликаты:**

2.1. Жанры игры (`genre`):

- Обнаружено: 12 неявных дубликатов из 24 уникальных значений

2.2. Платформы (`platform`):

- Обнаружено: 0 неявных дубликатов
- Уникальных значений: 31 (одинаково до и после нормализации)

2.3 Рейтинги (`rating`):

- Обнаружено: 0 неявных дубликатов
- Уникальных значений: 9 (включая 'Unknown' для пропущенных значений)

**3. Действия по обработке явных дубликатов:**

- Полные дубликаты: Удалены все полностью идентичные строки, оставлена первая встретившаяся запись

**4. Действия по обработке неявных дубликтов:**

4.1. Для жанров (`genre`):

- Приведены все значения к нижнему регистру 
- Устранены 12 неявных дубликатов

4.2. Для рейтингов (`rating`):

- Приведены все значения к верхнему регистру

4.3. Для платформ (`platform`):

- Неявных дубликатов не обнаружено, поэтому нормализация не требовалась

**Количество удалённых строк в абсолютном и относительном значениях.**

Изначально было 16956 строк, после удаления пропусков и явных дубликатов размер датафрема равен 16470. Следовательно было удалено 486 строк (абсолютное количество удаленных строк)

Относительное количество удаленных строк составляет: 2,86%

**Выводы**

1. Проведена проверка и нормализация названий столбцов. 

- Названия столбцов были приведены к единому формату snake_case

2. Проведен анализ и обработка типов данных.

- Выявлены некорректные типы в колонках продаж (object вместо float)

- Проведено преобразование типов данных

3. Проведена обработка пропущенных значений

- Удалены пропуски в `genre`, `name` и `year_of_release`

- Обработка пропусков в колонках `eu_sales` и `jp_sales` с помощью замены на среднее значение в зависимости от названия платформы и года выхода игры

- Обработка пропусков в колонках `critic_score` и `user_score` с помощью замены на значение медианы по платформе и году выпуска

- Пропуски в рейтингах заменены на категорию 'UNKNOWN'

4. Обнаружение и обработка неявных дубликатов

- **Колонка `genre`:** найдено 12 неявных дубликатов из-за разного регистра написания. Все значения приведены к нижнему регистру

- **Колонка `rating`:** значения переведены в верхний регистр, неявных дубликатов не обнаружено

- **Колонка `platform`**: проверена на опечатки, неявных дубликатов не обнаружено

5. Обнаружение и удаление явных дубликатов

- Обнаружены и удалены полные дубликаты (209 строк)

## 3. Фильтрация данных

In [41]:
# Создаем новый срез данных в отдельном датафрейме и отбираем данные по показателям
df_actual = df[(df['year_of_release'] >= 2000) & (df['year_of_release'] <= 2013)].copy()

In [42]:
# Выводим информацию о новом датафрейме
df_actual.info()

<class 'pandas.core.frame.DataFrame'>
Index: 12781 entries, 0 to 16954
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             12781 non-null  object 
 1   platform         12781 non-null  object 
 2   year_of_release  12781 non-null  Int16  
 3   genre            12781 non-null  object 
 4   na_sales         12781 non-null  float64
 5   eu_sales         12781 non-null  float64
 6   jp_sales         12781 non-null  float64
 7   other_sales      12781 non-null  float64
 8   critic_score     12781 non-null  float64
 9   user_score       12781 non-null  float64
 10  rating           12781 non-null  object 
dtypes: Int16(1), float64(6), object(4)
memory usage: 1.1+ MB


In [43]:
# Выводим первые строки нового датафрейма на экран
df_actual.head()

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
0,Wii Sports,Wii,2006,sports,41.36,28.96,3.77,8.45,76.0,8.0,E
2,Mario Kart Wii,Wii,2008,racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009,sports,15.61,10.93,3.28,2.95,80.0,8.0,E
6,New Super Mario Bros.,DS,2006,platform,11.28,9.14,6.50,2.88,89.0,8.5,E
7,Wii Play,Wii,2006,misc,13.96,9.18,2.93,2.84,58.0,6.6,E


## 4. Категоризация данных

In [44]:
# Создаеи копию датасета для работы с данными
df_actual = df_actual.copy()

In [45]:
# Создаем функцию для категоризации оценок пользователей
def categorize_user_score(score):

    if score >= 8 and score <= 10:
        return 'Высокая оценка'
    elif score >= 3 and score < 8:
        return 'Средняя оценка'
    elif score >= 0 and score < 3:
        return 'Низкая оценка'

# Создаем новую колонку с категориями
df_actual['user_score_category'] = df_actual['user_score'].apply(categorize_user_score)

In [46]:
# Создаем функцию для категоризации оценок критиков
def categorize_critic_score(score):

    if score >= 80 and score <= 100:
        return 'Высокая оценка'
    elif score >= 30 and score < 80:
        return 'Средняя оценка'
    elif score >= 0 and score < 30:
        return 'Низкая оценка'

# Создаем новую колонку с категориями
df_actual['critic_score_category'] = df_actual['critic_score'].apply(categorize_critic_score)

In [47]:
# Группировка данных по категориям оценок пользователей
groupby_user_score_category = df_actual.groupby('user_score_category')['user_score'].count()
display(groupby_user_score_category)

user_score_category
Высокая оценка    2286
Низкая оценка      116
Средняя оценка    4081
Name: user_score, dtype: int64

In [48]:
# Группировка данных по категориям оценок критиков
groupby_critic_score_category = df_actual.groupby('critic_score_category')['critic_score'].count()
display(groupby_critic_score_category)

critic_score_category
Высокая оценка    1692
Низкая оценка       55
Средняя оценка    5422
Name: critic_score, dtype: int64

In [49]:
# Подсчитываем количество игр для каждой платформы
platform_counts = df_actual['platform'].value_counts()

# Выбираем топ-7 платформ
top_7_platforms = platform_counts.head(7)

# Выводим результат
display(top_7_platforms)

platform
PS2     2127
DS      2120
Wii     1275
PSP     1180
X360    1121
PS3     1087
GBA      811
Name: count, dtype: int64

## 5. Итоговый вывод

Были загружены данные `new_games.csv`. Они содержат 11 столбцов и 16956 строк, в которых представлена информация о продажах компьютерных игр. В ходе проекта проведена комплексная предобработка данных и анализ игрового рынка за период 2000-2013 годы. Получены следующие результаты:

**1. Оптимизация типов данных**

- `year_of_release`: преобразован в Int64 для корректной работы с годами

- `eu_sales`, `jp_sales`: преобразованы из object в float64

 - `user_score`: преобразован из object в float64

**2. Первичный анализ и предобработка данных**

- Исходный датасет содержал 16956 записей о компьютерных играх с 11 характеристиками

- Выявлены и устранены проблемы с данными:

    - Пропуски в критических полях: Удалено 277 строк с пропусками в `name`, `genre` и `year_of_release`

    - Дубликаты: Удалено 209 полных дубликатов

    - Неявные дубликаты: Устранены 12 неявных дубликатов в жанрах из-за разного регистра

Итог очистки: Сохранено 97,14% от исходного объема (16470 строки), что обеспечивает репрезентативность выборки

**3. Создание актуального среза данных**

- Выделен период 2000-2013 годы

- Создан датафрейм df_actual с играми этого периода

- Объём среза: 12802 игр (около 75% от очищенных данных)

**4. Создание новых полей**

- Категоризация оценок пользователей (user_score_category):

    - Высокая оценка: 8-10 баллов (3084 игр)

    - Средняя оценка: 3-8 баллов (9597 игр)

    - Низкая оценка: 0-3 балла (121)

- Категоризация оценок критиков (critic_score_category):

    - Высокая оценка: 80-100 баллов (1809 игр)

    - Средняя оценка: 30-80 баллов (10938 игр)

    - Низкая оценка: 0-30 баллов (55 игр)

**5. Создание рейтинга**

- Выделены топ-7 платформ по количеству выпущенных игр

    - Лидеры: PS2, DS, Wii, PSP, X360, PS3, GBA

**Итог**

- Данные полностью очищены от ошибок и дубликатов

- Созданы аналитические категории для упрощения анализа

- Выделен репрезентативный временной период

- Данные структурированы и готовы для дальнейшего изучения